tests/PHPStan/Rules/Methods/data/bug-8523b.php

```diff
@@ -0,0 +1,24 @@
+<?php declare(strict_types = 1);
+
+namespace Bug8523b;
+
+class HelloWorld
+{
+	public static ?HelloWorld $instance = null;
+
+	public function save(): void {
+		self::$instance = new HelloWorld();
+
+		$callback = static function(): void {
+			self::$instance = null;
+		};
+
+		$callback();
+
+		var_dump(self::$instance);
+
+		self::$instance?->save();
+	}
+}
+
+(new HelloWorld())->save();
```

Comment:
there is a similar case where we still fail:

```
<?php declare(strict_types = 1);

class HelloWorld
{
	public static ?HelloWorld $instance = null;
	
    public function save(): void {
        HelloWorld::$instance = new HelloWorld();

        $callback = static function(): void {
			HelloWorld::$instance = null;
		};
		
		$callback();
			
		var_dump(HelloWorld::$instance);
		
        HelloWorld::$instance?->save();
    }
}

(new HelloWorld())->save();
```

note: instead of `self` in this case the FQCN is used.

maybe we should normalize FQCN in AST to `self` in such cases, so we don't need to handle this case all over the place?